src/Interpreters/SessionLog.cpp

```diff
@@ -0,0 +1,172 @@
+}
+
+using AuthType = Authentication::Type;
+using Interface = ClientInfo::Interface;
+
+void fillColumnArray(const Strings & data, IColumn & column)
+{
+    auto & array = typeid_cast<ColumnArray &>(column);
+    size_t size = 0;
+    for (const auto & name : data)
+    {
+        array.getData().insertData(name.data(), name.size());
+        ++size;
+    }
+    auto & offsets = array.getOffsets();
+    offsets.push_back(offsets.back() + size);
+};
+
+}
+
+namespace DB
+{
+
+SessionLogElement::SessionLogElement(Type type_, String user_)
+    : type(type_),
+      user(std::move(user_))
+{
+    std::tie(event_time, event_time_microseconds) = eventTime();
+}
+
+Block SessionLogElement::createBlock()
+{
+    auto event_type = std::make_shared<DataTypeEnum8>(
+        DataTypeEnum8::Values
+        {
+            {"LoginFailure",           static_cast<Int8>(SESSION_LOGIN_FAILURE)},
+            {"LoginSuccess",           static_cast<Int8>(SESSION_LOGIN_SUCCESS)},
+            {"Logout",                 static_cast<Int8>(SESSION_LOGOUT)}
+        });
+
+    auto identified_with_column = std::make_shared<DataTypeEnum8>(
+        DataTypeEnum8::Values
+        {
+            {"NO_PASSWORD",            static_cast<Int8>(AuthType::NO_PASSWORD)},
+            {"PLAINTEXT_PASSWORD",     static_cast<Int8>(AuthType::PLAINTEXT_PASSWORD)},
+            {"SHA256_PASSWORD",        static_cast<Int8>(AuthType::SHA256_PASSWORD)},
+            {"DOUBLE_SHA1_PASSWORD",   static_cast<Int8>(AuthType::DOUBLE_SHA1_PASSWORD)},
+            {"LDAP",                   static_cast<Int8>(AuthType::LDAP)},
+            {"KERBEROS",               static_cast<Int8>(AuthType::KERBEROS)}
+        });
+
+    auto interface_type_column = std::make_shared<DataTypeEnum8>(
+        DataTypeEnum8::Values
+        {
+            {"TCP",                    static_cast<Int8>(Interface::TCP)},
+            {"HTTP",                   static_cast<Int8>(Interface::HTTP)},
+            {"gRPC",                   static_cast<Int8>(Interface::GRPC)},
+            {"MySQL",                  static_cast<Int8>(Interface::MYSQL)},
+            {"PostgreSQL",             static_cast<Int8>(Interface::POSTGRESQL)}
+        });
+
+    return
+    {
+        {std::move(event_type),                   "type"},
+        {std::make_shared<DataTypeDate>(),        "event_date"},
+        {std::make_shared<DataTypeDateTime>(),    "event_time"},
+        {std::make_shared<DataTypeDateTime64>(6), "event_time_microseconds"},
+
+        {std::make_shared<DataTypeString>(),      "user"},
+        {std::move(identified_with_column),       "authenticated_by"},
+        {std::make_shared<DataTypeArray>(
+            std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>())), "profiles"},
+        {std::make_shared<DataTypeArray>(
+            std::make_shared<DataTypeLowCardinality>(std::make_shared<DataTypeString>())), "roles"},
```

Comment:
There are two kinds of active roles. `Current roles` are the roles which are chosen to be used by user. `Enabled roles` includes
1) all the `current roles`, and
2) all the roles which are granted to the `current roles`, and
3) all the roles which are granted to `2`, and
4) all the roles which are granted to `3` and so on.

I mean maybe we want to see the current roles in the `session_log` table, not the enabled ones, what do you think?

Reply:
Ok, makes sense